search

kedashoe opened the pull request:
Supersedes #11. Now using hindley milner notation for the search input. Things seem to be working pretty well?

I've brought in `browserify` to tie in external packages and `uglify` to minify. I don't really know anything about minification, so if anyone has some thoughts there I am happy to go with something else for that step.

One kinda funny thing: you can't search for combinators by name because the input parser does not allow function names (correctly, I believe?) to start with upper case letters. We could probably create an exception for them.

Todo:

- [x] lint `src/search.js`
- [x] properly index type variables (so that eg `x -> x` matches `a -> a`)
- [x] fuzzy matching too many results?
- [ ] update query string at __/search/__ as input changes
- [ ] parse query string to populate search field
- [ ] add search field to __/__ which redirects to __/search/?q=…__

kedashoe committed:
search

davidchambers commented:
Very cool, Kevin!

I'm sure I'll have more thoughts, but these are my initial ones:

  - When searching, I expect the list of results to contain only matches and their ancestors. "Overview", for example, need never appear in the list of results.

  - `Either a b` produces many matches, but `Either x y` produces none. This suggest that we're not treating type variables as variables. `a -> a` and `x -> x` should both match `S.I`.

  - `Maybe#map` produces many matches for some reason.

kedashoe commented:
> When searching, I expect the list of results to contain only matches and their ancestors. "Overview", for example, need never appear in the list of results.

makes sense

> Either a b produces many matches, but Either x y produces none. This suggest that we're not treating type variables as variables. a -> a and x -> x should both match S.I

ya, should have mentioned this one

> Maybe#map produces many matches for some reason.

I think I am letting the fuzzy searcher be too generous? Will play around with it.

davidchambers commented:
Having given the matter more thought I now believe the search functionality should have its own URL:

    https://sanctuary.js.org/search/

We could then share links to search results which would be useful. For example:

    https://sanctuary.js.org/search/?q=List%20a%20-%3E%20Maybe%20a

We could make this discoverable by including a text input near the top of the main page (as you have currently), and setting `window.location` to the appropriate value when a query is submitted.

kedashoe commented:
Pushed an update. The search now manages its own list rather than showing/hiding items in the ToC (didn't occur to me at first, but this was always going to be necessary). `x -> Maybe y` should now return appropriate results. Do you want the URL stuff to be part of this PR?

davidchambers commented:
> Pushed an update.

Looks good!

> `x -> Maybe y` should now return appropriate results.

I see these results:

  - `Maybe#ap :: Maybe (a -> b) ~> Maybe a -> Maybe b`
  - `Maybe#chain :: Maybe a ~> (a -> Maybe b) -> Maybe b`
  - __`Maybe#equals :: Maybe a ~> b -> Boolean`__
  - `Maybe#of :: Maybe a ~> b -> Maybe b`
  - __`Maybe#reduce :: Maybe a ~> ((b, a) -> b) -> b -> b`__
  - `maybe :: b -> (a -> b) -> Maybe a -> b`
  - `encase :: (a -> b) -> a -> Maybe b`
  - `maybeToEither :: a -> Maybe b -> Either a b`
  - `eitherToMaybe :: Either a b -> Maybe b`

The results in bold seem out of place to me.

> Do you want the URL stuff to be part of this PR?

Yes please! Let me know if I can help with this in any way. :)

kedashoe commented:
ah, forgot to mention that. In order to allow us to find functions and methods, I am compiling methods to their function equivalent. So

:: Maybe a ~> b -> Boolean

is compiled to

:: t0 -> Maybe t1 -> Boolean

https://github.com/kedashoe/hindley-milner-search/blob/master/src/hm-search.js#L68-L78

davidchambers commented:
> I am compiling methods to their function equivalent

It might be nice for `~>` to return all methods, and for `Maybe a ~>` to return all Maybe methods. It's worth reconsidering your rewrite rule.

By the way, I'm quite confused by the results that `~>` currently returns!

kedashoe commented:
> It might be nice for ~> to return all methods, and for Maybe a ~> to return all Maybe methods. It's worth reconsidering your rewrite rule.

I had thought we were trying to discourage the use of methods? The difficulty is that I think we definitely want `Maybe a -> b` to return methods if they match, but it would not be straightforward to be able to match both function syntax and the `Maybe a ~> b` syntax (off the top of my head, every signature would need to have multiple parse trees which would all need to be searched).

> By the way, I'm quite confused by the results that ~> currently returns!

Anything that we cannot parse we treat as a name, and `~>` doesn't parse to anything, so it falls back to a name. Those signatures that it matches are also ones we cannot (currently) parse, so they also become names, and the fuzzy matcher likes those best. The fuzzy matcher returns a score, maybe if everything scores very poorly I should say "No matches"?

davidchambers commented:
This discussion has made it apparent to me that fuzzy matching algorithms don't sit well with me. Web search engines such as Google are incredibly useful, of course, and are necessarily fuzzy. In more limited domains, though, I prefer algorithms which I understand (or could understand given enough time). I would like to be able to state definitively—without consulting a computer—the matches for a given search term.

I'm quite possibly anomalous.

I don't want my preferences to block this useful feature. Let's get the user interface in good shape then release the feature so that people can start using it and provide feedback. We can tweak the weightings as we see fit.

kedashoe commented:
I don't have any attachment to the fuzzy search. I was planning to write a simpler one at some point anyway. For now, I'll play around with the scores it returns, I think returning some results no matter what is causing more harm than good.

kedashoe commented:
I've updated the fuzzy searcher. No third party lib, just very simple search and score. Results look pretty good to me. I've also removed type info from sigs that can't be parsed, this seemed to be causing some confusion.

Also rebased. The bold that was added looks really nice in the search results :)

When you are ready to merge, please post to let me know before merging. I will release 1.0.0 of both the parser and searcher at that point, update the deps, make sure everything is still working, and then let you know it is ready to merge.

davidchambers commented on package.json:
```diff
@@ -10,9 +10,12 @@
     "marked": "0.3.5",
-    "sanctuary-style": "0.4.x"
+    "sanctuary-style": "0.4.x",
+    "uglify-js": "^2.7.5"
```
Is there a reason to use `^2.7.5` rather than `2.7.x` in this case?

davidchambers commented on src/search.js:
```diff
@@ -0,0 +1,125 @@
+  return function(el) {
+    container.appendChild(el);
+  };
+}
```
Since we have Sanctuary available to us, should we use `S.curry2` here?

davidchambers commented on src/search.js:
```diff
@@ -0,0 +1,125 @@
+  var codes = document.getElementById('toc').querySelectorAll('li a code');
+  var as = map(getParent, codes);
+  var newEls = map(initSigResult, as);
+  return newEls;
```
I don't think the names add clarity here. Let's nest expressions:

```javascript
return map(initSigResult, map(getParent, codes));
```

svozza replied:
Do we have `compose` so we don't map over it twice:

```js
map(compose(initSigResult, getParent), codes)
```

davidchambers commented on src/search.js:
```diff
@@ -0,0 +1,125 @@
+
+// Node -> String
+function extractSignatureText(el) {
+  return el.innerText.trim().replace(/\s+/g, ' ');
```
`S.unwords(S.words(el.innerText))` would be nice here. :)

davidchambers commented on src/search.js:
```diff
@@ -0,0 +1,125 @@
+  }
+  else {
+    hide(state.containerNo);
+  }
```
I don't much like the `show` and `hide` functions as they're impure. I'd prefer this mutation to appear in situ:

```javascript
state.containerNo.style.display =
  matches.length === 0 && q.length > 0 ? 'block' : 'none';
```

davidchambers commented on src/search.js:
```diff
@@ -0,0 +1,125 @@
+  clearChildren(state.container);
+  // if there is no query, do not show any matches
+  if (q.length > 0) {
+    for (var i = 0; i < matches.length; ++i) {
```
Let's use `matches.forEach` here to avoid the index variable.

davidchambers commented on src/search.js:
```diff
@@ -0,0 +1,125 @@
+  searchInput.focus();
+  searchInput.addEventListener('keyup', handleSearchKeyUp);
+});
+
```
:space_invader:

davidchambers commented on style.css:
```diff
@@ -111,6 +111,10 @@ a code {
   top: -0.05em;
 }
 
+#search-results > div {
```
Matching elements by tag name is inadvisable. Let's add an id or class name to the div and use a more precise selector here.

davidchambers commented on src/search.js:
```diff
@@ -0,0 +1,125 @@
+  return Array.prototype.map.call(x, f);
+}
+
+// Node -> Node
```
I'd like to be able to run `ack '^ *//' src/search.js` to view all the type signatures. Please update this signature (and others) to include the function name:

```javascript
//  getGrandparent :: Node -> Node
```

kedashoe commented:
updated.

My eslint knowledge is next to zero. I've fixed all the ones that looked like they should be fixed and am left with:

- error  'require' is not defined
- error  Use the function form of 'use strict'
- error  Expected a function declaration

On that last one, since I am using `S` now and currying a few functions, I changed all function declarations to be `var foo = function foo...` to be consistent.

davidchambers commented:
Are you happy for me to push to your branch, Kevin? I'd like to have a go at adding a dedicated __/search/__ page.

kedashoe commented:
Sure, do I need to do anything?

davidchambers commented:
> Sure, do I need to do anything?

Nope. I'm going to rebase to resolve merge conflicts before I make any changes. Once I've done so you'll need to check out a fresh copy of this branch to avoid merge conflicts locally.

davidchambers committed:
refactoring

davidchambers commented:
We must resolve kedashoe/hindley-milner-search#4 before merging this pull request.

davidchambers committed:
hindley-milner-search@0.1.3

davidchambers commented:
Do we have any control over the fuzziness? It's great that a search for `map` finds <code><b>map</b></code>, <code>bi<b>map</b></code>, and <code>pro<b>map</b></code>, but I don't think it should find <code><b>Ma</b>ybe#ins<b>p</b>ect</code>.

kedashoe commented:
> Do we have any control over the fuzziness? It's great that a search for map finds map, bimap, and promap, but I don't think it should find Maybe#inspect.

see https://github.com/kedashoe/hindley-milner-search/pull/6. will merge and release if that looks good to you.

kedashoe commented:
> will merge and release if that looks good to you.

0.2.0 up

davidchambers committed:
hindley-milner-search@0.2.x

davidchambers commented:
I've updated this branch to `hindley-milner-search@0.2.x`. We're now specifying `{fuzzy: false}` which fixed the overeager matching of names, but now a search for `a -> a` returns no results.

@kedashoe, can we perform substring matching of function names without losing the ability to search by type?

kedashoe commented:
Sorry about that, I've released v0.2.1 with the fix.

davidchambers committed:
hindley-milner-search@0.2.1

davidchambers commented:
Thanks for the update, Kevin. :)

I still can't find `S.I`: neither `I` nor `a -> a` finds it. Have you any idea why this is the case?

kedashoe commented on src/search.js:
```diff
@@ -0,0 +1,58 @@
+    var search = HMS.init(Array.prototype.map.call($els, function(el) {
+      return el.innerText.trim().split(/\s+/).join(' ');
+    }), {fuzzy: false}).search;
+    var $prevSearch = '';
```
What does `$` prefix mean? Why does `$prevSearch` have one but not `q` below?

davidchambers replied:
In https://github.com/sanctuary-js/sanctuary-site/commit/6e24452b90e8f4894dbc9e590ed5e51c18795199 I replaced the `state` object with various variables. I assumed each of these variables to be stateful, and used the `$` prefix to remind myself of this.

kedashoe commented on src/search.js:
```diff
@@ -0,0 +1,58 @@
+          $container.removeChild($container.childNodes[0]);
+        }
+        // if there is no query, do not show any matches
+        if (q !== '') {
```
I originally had the empty string return all matches simply because I was culling results from the current DOM rather than building a new list. I'm not sure which would be more common, but I can add an option to return all or none on empty input and we wouldn't need this conditional in userland either way.

davidchambers replied:
Since we'll be displaying results at __/search/__ rather than at __/__, displaying all results for an empty query seems right to me. I've removed this check. :)

kedashoe replied:
I had to move this logic a little in the latest release. Not a million percent sure we're on the same page but just let me know if empty string is not doing what you want now.

davidchambers committed:
show everything when searching for ""

davidchambers committed:
remove $prevSearch optimization

davidchambers committed:
add search/index.html

kedashoe commented:
New release up, `a -> a` and `I` should both find `I` now.

One other note, when I was playing around with my local copy of sanctuary-site, I was not getting many results for anything at first. This was due to `\u2060` in our arrows (I feel like I remember discussion about this but couldn't find the PR), removing those fixed things.

davidchambers committed:
hindley-milner-search@0.2.2

davidchambers committed:
move search functionality to /search/

davidchambers committed:
remove unused file

davidchambers commented on Makefile:
```diff
@@ -22,6 +23,12 @@ vendor/ramda.js: node_modules/ramda/dist/ramda.js
 	cp '$<' '$@'
 
+js/hm-search.js: node_modules/hindley-milner-search/hm-search.js Makefile
+	$(BROWSERIFY) --standalone HMS -- '$<' >'$@'
```
It would be nice if a browser build were available. This is related to kedashoe/hindley-milner-parser-js#1.

kedashoe replied:
browser standalone files included in v0.2.3. You should see `hms.js` and `hms.min.js` now. Exported as `HMS` as you did.

davidchambers replied:
Excellent! I've updated the makefile to simply copy __hms.js__, so we no longer depend on Browserify. :)

davidchambers committed:
hindley-milner-search@0.2.3

chughes87 commented:
Bump on this PR! I'd love to see this merged soon...
